Bottom: 4968c3363e56411ef336cdfdb67f27624761f7eb
Top:    67e9c6c54cb90a8d2cbd3353e2ee6edd5c36d6d6
Author: Duncan Mac-Vicar P <dmacvicar@suse.de>
Date:   2013-03-06 16:40:33 +0100




---

diff --git a/createrepo/__init__.py b/createrepo/__init__.py
index 8f2538e..69610f6 100644
--- a/createrepo/__init__.py
+++ b/createrepo/__init__.py
@@ -24,6 +24,7 @@ from  bz2 import BZ2File
 from urlgrabber import grabber
 import tempfile
 import stat
+import re
 import fcntl
 import subprocess
 
@@ -504,6 +505,30 @@ class MetaDataGenerator:
                            " add it" % po
 
         return po
+ 
+    def sortPackages(self, pkglist):
+        class PackageKey:
+            def __init__(self, package):
+                m = re.search(r"([^/]+)-([^/-]+-[^/-]+)\.([^/.]+)\.rpm$",
+                        package)
+                if m:
+                    self.n = m.group(1)
+                    self.vr = m.group(2)
+                    self.a = m.group(3)
+                else:
+                    # give up on sorting
+                    self.n = package
+                    self.vr = 0
+                    self.a = 0
+            def __lt__(self, other):
+                if self.n != other.n:
+                    return self.n < other.n
+                if self.a != other.a:
+                    return self.a < other.a
+                return self.vr < other.vr
+
+        return sorted(pkglist, key=PackageKey)
+
 
     def writeMetadataDocs(self, pkglist=[], pkgpath=None):
 
@@ -515,6 +540,8 @@ class MetaDataGenerator:
         else:
             directory = pkgpath
 
+        pkglist = self.sortPackages(pkglist)
+
         # for worker/forked model
         # iterate the pkglist - see which ones are handled by --update and let them
         # go on their merry way
